src/sdk/pynni/nni/compression/torch/utils/shape_dependency.py

```diff
@@ -0,0 +1,191 @@
+            # user should provide model & dummy_input to trace the model or a already traced model
+            assert model is not None and dummy_input is not None
+        self.graph = TorchModuleGraph(model, dummy_input, traced_model)
+        self.dependency = dict()
+        self.build_channel_dependency()
+
+    def get_parent_convs(self, node):
+        """
+        Find the nearest father conv layers for the target node.
+
+        Parameters
+        ---------
+        node : torch._C.Node
+            target node.
+
+        Returns
+        -------
+        parent_convs: list
+            nearest father conv layers for the target worknode.
+        """
+        parent_convs = []
+        queue = []
+        queue.append(node)
+        while queue:
+            curnode = queue.pop(0)
+            if curnode.op_type == 'Conv2d':
+                # find the first met conv
+                parent_convs.append(curnode.name)
+                continue
+            parents = self.graph.find_predecessors(curnode.unique_name)
+            parents = [self.graph.name_to_node[name] for name in parents]
+            for parent in parents:
+                queue.append(parent)
+        return parent_convs
+
+    def build_channel_dependency(self):
+        """
+        Build the channel dependency for the conv layers
+        in the model.
+        """
+        for node in self.graph.nodes_py.nodes_op:
+            parent_convs = []
+            # find the node that contains aten::add
+            # or aten::cat operations
+            if node.op_type in ADD_TYPES:
+                parent_convs = self.get_parent_convs(node)
+            elif node.op_type == CAT_TYPE:
+                # To determine if this cat operation will introduce channel
+                # dependency, we need the specific input parameters of the cat
+                # opertion. To get the input parameters of the cat opertion, we
+                # need to traverse all the cpp_nodes included by this NodePyGroup,
+                # because, TorchModuleGraph merges the important nodes and the adjacent
+                # unimportant nodes (nodes started with prim::attr, for example) into a
+                # NodepyGroup.
+                cat_dim = None
+                for cnode in node.node_cpps:
+                    if cnode.kind() == CAT_TYPE:
+                        cat_dim = list(cnode.inputs())[1].toIValue()
+                        break
+                if cat_dim != 1:
+                    parent_convs = self.get_parent_convs(node)
+            dependency_set = set(parent_convs)
+            # merge the dependencies
+            for parent in parent_convs:
+                if parent in self.dependency:
+                    dependency_set.update(self.dependency[parent])
+            # save the dependencies
+            for _node in dependency_set:
+                self.dependency[_node] = dependency_set
+
+
+    def filter_prune_check(self, ratios):
```

Comment:
how this function is supposed to be used? by pruner?

Reply:
I think the pruner may need such an interface, if not, we can delete this function.

Reply:
yes, we can add it back when a pruner uses this function.

Reply:
do we support linear layers?